[WIP] Adding PPC plot to Arviz-Plots

imperorrp opened the pull request:
Adding plot_ppc (issue https://github.com/arviz-devs/arviz-plots/issues/11). 

This is a PPC plot draft that currently plots the actual observed values and prior/posterior predictive values, flattening across all dimensions except sample dims (chain, draw) by default 

### WIP: 
- The prior/posterior predictive mean
- Cumulative and scatter visualization 

<!-- readthedocs-preview arviz-plots start -->
----
📚 Documentation preview 📚: https://arviz-plots--55.org.readthedocs.build/en/55/

<!-- readthedocs-preview arviz-plots end -->

imperorrp commented:
Plots generated by passing in the `centered_eight` datatree as `data`:

It seems like multiple colors are cycled through even though no aesthetic mapping was defined in kwargs for the kde curves for the ppc samples so the observed values curve isn't standing out. Not sure how to fix this-

`azp.plot_ppc(data, data_pairs={"y":"y"}, num_pp_samples=50)`
![image](https://github.com/arviz-devs/arviz-plots/assets/24991131/2473f9e5-a8cd-4b1e-9fe3-835dde5fe1c3)

`azp.plot_ppc(data, data_pairs={"y":"y"}, num_pp_samples=500)`
![image](https://github.com/arviz-devs/arviz-plots/assets/24991131/64d8cb88-7ca9-4380-b5c5-944b9e7e152d)

When only 1 pp sample is selected though, the observed curve becomes clear (the darker one)-
`azp.plot_ppc(data, data_pairs={"y":"y"}, num_pp_samples=1)`
![image](https://github.com/arviz-devs/arviz-plots/assets/24991131/f14641ee-e029-4d1c-b9f8-2a3322a433e5)

As a comparison for checking the visualization accuracy, this is the plot generated for the same `centered_eight` data by the legacy Arviz plot_ppc-
![image](https://github.com/arviz-devs/arviz-plots/assets/24991131/c0822417-d424-4b55-8d94-001364f3a476)

imperorrp commented on src/arviz_plots/plots/ppcplot.py:
```diff
@@ -0,0 +1,385 @@
+    pp_densities = []
+    pp_xs = []
+
+    for i in range(predictive_distribution.sizes["ppc_dim"]):
```
So this is sort the approach that should be taken when calling map on plot_ppc as well, right - as in sample_dims should be all of the non chain and draw dimensions (by default, though a user could select less) since these are the dims being flattened and used to generate KDEs, while the actual pp sample dims chain and draw are set as "overlay" in the aesthetic mapping?

OriolAbril replied:
In the example `sample_dims` has nothing to do with `.map`, the need for `sample_dims="school"` is because `sample_dims` are the dimensions being reduced by `plot_dist` when computing the kde.

In plot_ppc, we split the dimensions into 3 non-overlapping groups:
* overlay dimensions -> here `sample_dims`
* facet dimensions -> what used to be `flatten` argument/now modifiable also via pc_kwargs
* rest/leftover/reduce (defined by elimination) -> these are the ones that should be reduced when computing the kde

In this particular example, they are respectively `[chain, draw]`, `[]` and `[school]`

imperorrp replied:
Got it, this way of splitting makes sense. I've modified the code in my latest commit to take into account these three kinds of dimensions (for overlay/facetting/reducing)

imperorrp commented:
Latest commit takes into account your suggestions and comments @OriolAbril. The `mean` plotting might need some more work as the current implementation just takes into account all of the data to generate the mean. If the plot is facetted, generating the mean curves on the relevant subselections of the data would then be required I suppose.

Here is the current output when plot_ppc is called like:
```
azp.plot_ppc(data, num_pp_samples=500)
```
![image](https://github.com/arviz-devs/arviz-plots/assets/24991131/2af0cfa3-743b-4d51-bf05-ce1e4d4f4be7)

OriolAbril commented on src/arviz_plots/plots/ppcplot.py:
```diff
@@ -0,0 +1,479 @@
+            kind=kind,
+            plot_collection=plot_collection,
+            labeller=labeller,
+            aes_map={
+                kind: value
+                for key, value in aes_map.items()
+                if key == "predictive" and key not in pp_density_ignore
+            },
```
I don't understand what is happening here. The keys of the `aes_map` dict need to be ids (aka artist names) for the visual elements in plot_dist, the values are sets of aesthetic names. It indicates which aesthetics should be applied to each artist.

No key can be in `pp_density_ignore` as `pp_density_ignore` contains aesthetic names. That means only the "predictive" key will be used, which makes sense as it is what we want to plot, and it is "renamed" to `kind`, but also `kind` is hardcoded even if inside the dict comprehension, so if the loop were to actually do anything then it would break as it would attempt to generate a dictionary with repeated keys.

imperorrp replied:
I think I meant to put `if key == "predictive" and value not in pp_density_ignore` so it checks the aesthetic names against those in `pp_density_ignore`. About the dist comprehension yeah it doesn't seem to useful in retrospect. I'll refactor this to how it's done in `plot_trace_dist` when `plot_dist` is called intrinsically

OriolAbril commented on src/arviz_plots/plots/ppcplot.py:
```diff
@@ -0,0 +1,479 @@
+                **{key: False for key in plot_dist_artists},
+                kind: dict(pp_kwargs.items()),
+            },
+            stats_kwargs={key: value for key, value in stats_kwargs.items() if key == "predictive"},
```
Somewhat similar comment here. Make sure the code makes sense and that it is consistent with the documentation.

OriolAbril commented on src/arviz_plots/plots/ppcplot.py:
```diff
@@ -0,0 +1,479 @@
+    # ---------STEP 2 (PPC AGGREGATE)-------------
+
+    aggregate_kwargs = copy(plot_kwargs.get("aggregate", {}))
+    if aggregate and pp_kwargs is not False and aggregate_kwargs is not False:
```
This should not depend on `pp_kwargs`, only on `aggregate` related kwargs. Also, this should be consistent with whatever happens with `observed` and `plot_kwargs["observed"]` if one raises a warning when there is duplication, the other should too, if it is an error instead, both should raise an error (ideally if an error at the top of the function before any computation or plotting happens).

OriolAbril commented on src/arviz_plots/plots/ppcplot.py:
```diff
@@ -0,0 +1,479 @@
+        aggregate_reduce_dims = reduce_dims + list(sample_dims)
+        print(f"\n aggregate_reduce_dims = {aggregate_reduce_dims}")
+
+        # setting aggregate aes_map to `[]` so `overlay` isn't applied
+        aes_map.setdefault("aggregate", [])
```
This should happen in the aes_map defaults (it doesn't really matter if we set that and then aggregate=False)

OriolAbril commented on src/arviz_plots/plots/ppcplot.py:
```diff
@@ -0,0 +1,479 @@
+        )
+        obs_kwargs = copy(plot_kwargs.get("observed", {}))
+
+        if obs_kwargs is not False:
```
This still has the double level logic with `observed` and `obs_kwargs`

OriolAbril commented on src/arviz_plots/plots/ppcplot.py:
```diff
@@ -0,0 +1,479 @@
+
+    if stats_kwargs is None:
+        stats_kwargs = {}
+    if facet_dims is None:
+        facet_dims = []
+
+    # check for duplication of facet_dims in top level arg input and pc_kwargs
+    if "cols" in pc_kwargs and len(facet_dims) > 0:
+        duplicated_dims = set(facet_dims).intersection(pc_kwargs["cols"])
+        with warnings.catch_warnings():
+            warnings.filterwarnings(
+                "ignore",
+                message=f"""Facet dimensions have been defined twice.
+                Both in the the top level function arguments and in `pc_kwargs`. 
+                The `cols` key in `pc_kwargs` will take precedence.
+                                        
+                facet_dims = {facet_dims}
+                pc_kwargs["cols"] = {pc_kwargs["cols"]}
+                Duplicated dimensions: {duplicated_dims}.""",
+            )
+        # setting facet_dims to pc_kwargs defined values since it is used
+        # later to calculate dims to reduce
+        facet_dims = list(set(pc_kwargs["col"]).difference({"__variable__"}))
```
Extra note: reviewing the plot_ridge PR, I realized that a similar thing already happens in `plot_forest` and we raise errors, so for consistency I think we should raise errors here too. ref https://github.com/arviz-devs/arviz-plots/blob/main/src/arviz_plots/plots/forestplot.py#L242-L249

imperorrp replied:
Ah my bad, I just used the error bit from `plot_dist` as a template without checking it properly. I'll replace with the `raise ValueError(...)` you pointed out from `plot_forest`. 

Also, I think the first option sounds good- raise an error if both are provided and ask the user to only provide one of the args.

imperorrp commented on src/arviz_plots/plots/ppcplot.py:
```diff
@@ -0,0 +1,479 @@
+    if observed is None:
+        observed = group == "posterior"  # by default true if posterior, false if prior
+
+    # checking if plot_kwargs["observed"] or plot_kwargs["aggregate"] is not inconsistent with
+    # top level bool arguments `observed`` and `aggregate`
+    if "observed" in plot_kwargs:
+        if plot_kwargs["observed"] != observed:
+            with warnings.catch_warnings():
+                warnings.filterwarnings(
+                    "ignore",
+                    message="""`plot_kwargs['observed']` inconsistency detected.
+                    It is not the same as with the top level `observed` argument. 
+                    `plot_kwargs['observed']` will take precedence.""",
+                )
+            observed = plot_kwargs["observed"]
```
I'm a bit confused by this logic (Wouldn't this imply if both are true, the error is raised?) Did you mean 
```if plot_kwargs.get("observed", False) is not False and not observed:```? (implying `plot_kwargs["observed"]` is set to true but `observed` is false, so an error should be raised as they contradict)

Also what about the condition when `observed` is true but `plot_kwargs["observed"]` is false- should we display the error for this condition too due to contradiction, or just let the top-level argument `observed` take precedence? 

Because it might be the case that `observed` is not set at all by the user but defaults to true (which happens in the code before this check) because of the posterior predictive group being plotted, but the user has set `plot_kwargs["observed"]` to false to not have it be displayed- in this case should the top level argument still take preference even though it took a default value and wasn't explicitly defined in the function call?

OriolAbril replied:
You are right, the code makes no sense, sorry. I started writing with an idea on mind on how to do it then changed my mind but not the text.

The two options should be considered:

```python
if (
    observed and plot_kwargs.get("observed", True) is False
    or not observed and plot_kwargs.get("observed", False) is not False
):
    # raise warning or error
```

OriolAbril replied:
Whenever they are incompatible we should raise an error, thus, none will ever take precedence over the other because they either mean the same thing or an error is thrown

OriolAbril replied:
Actually, depending on #66 we cpuld simplify it to " `plot_kwargs["observed"] ` can't be False, use observed top level kwarg for that"

imperorrp replied:
I've used the first incompatibility logic for now- could we keep a note of this and make modifications after the proposals in #66 are finalized?

imperorrp commented:
Made some modifications/fixes. Some changes related to input kwargs (the incompatibility checks between top level args and `plot_kwargs` keys) might have to be further modified later though (in relation to issue #66) in the future. 

Updated the docstring as well so users are aware that the values they pass into `plot_ppc` related keys ("predictive", "aggregate", and "observed") in `plot_kwargs` and `aes_map` are transferred to `plot_dist`'s 'kind' type artists internally. In the code for this, I've used `plot_trace_dist` conventions as far as possible but with some changes since `plot_dist` is only called once in that plot and not for multiple artists.

`stats_kwargs` is passed to all internally called `plot_dist`s as is- so there is an assumption being made that users probably will want the same kind of statistical computation run for each `plot_ppc` artist they want to plot for consistency.

OriolAbril commented:
I'll try to review the code tomorrow, in the meantime some quick amswers to the questions above.

I don't think it matters to users that plot_dist is called internally, they only care about the valid keys and where are their values passed to eventually (so they know what values are valid). This will also match how things are documented in plot_trace_dist.

stats_kwargs should allow for different kwargs being passed to the different datasets, see https://github.com/arviz-devs/arviz-plots/pull/55#discussion_r1648995395

imperorrp commented:
> I'll try to review the code tomorrow, in the meantime some quick amswers to the questions above.
> 
> I don't think it matters to users that plot_dist is called internally, they only care about the valid keys and where are their values passed to eventually (so they know what values are valid). This will also match how things are documented in plot_trace_dist.
> 
> stats_kwargs should allow for different kwargs being passed to the different datasets, see [#55 (comment)](https://github.com/arviz-devs/arviz-plots/pull/55#discussion_r1648995395)

Okay, I'll adjust the docstring to reflect that so its simpler- I could make mention of it as a comment in the code though so if looking at the source code one could make out where it's being passed internally. 

Also sorry, missed that. I'll adjust the stats_kwargs usage logic to allow different ones for "predictive", "aggregate" and "observed".

imperorrp commented:
Added observed_rug to plot the observed values' distribution. Since no masking is required as divergences aren't being plotted here, the `trace_rug` visual element was also modified a bit to allow for mask=None. 

![image](https://github.com/user-attachments/assets/afb9560e-6410-401a-b6ef-b9516a5e767a)

OriolAbril commented on src/arviz_plots/plots/ppcplot.py:
```diff
@@ -23,16 +23,14 @@ def plot_ppc(
     sample_dims=None,
     kind=None,
     facet_dims=None,
-    data_pairs=None,
```
data_pairs argument should be kept

OriolAbril commented on src/arviz_plots/plots/ppcplot.py:
```diff
@@ -0,0 +1,507 @@
+
+    # pp distribution group plotting logic
+    pp_distribution = process_group_variables_coords(
+        dt, group=predictive_data_group, var_names=var_names, filter_vars=filter_vars, coords=coords
```
`data_pairs` should be taken into account here:

```suggestion
        dt, group=predictive_data_group, var_names=None if var_names is None else [data_pairs.get(var_name, var_name) for var_name in var_names], filter_vars=filter_vars, coords=coords
```

OriolAbril reviewed (commented):
This needs tests so we can check the code does what we want to happen in all the different possible scenarios

imperorrp commented:
About the tests- the 'kind' and 'group' arguments I've parametrized both via pytest and via hypothesis.strategies. Should only one be picked or this kept? 

The hypothesis tests are also failing a few times. At least one seems to be because of trying to toggle the 'observed' artist off with plot_kwargs but not in the top level arg, so the ValueError set for this gets raised. The others I am still unsure of why they're failing.

OriolAbril commented on src/arviz_plots/plots/ppcplot.py:
```diff
@@ -0,0 +1,529 @@
+        Dimensions to loop over in plotting posterior/prior predictive.
+        Note: Dims not in sample_dims or facet_dims (below) will be reduced by default.
+        Defaults to ``rcParams["data.sample_dims"]``
+    kind : {"kde", "cumulative", "scatter"}, optional
```
```suggestion
    kind : {"kde", "cumulative", "scatter"}, optional
```

I see you did some changes like renaming "cumulative" to "ecdf" internally, but that will be confusing for users. There should be 4 possible values accepted by `kind`: kde, ecdf, hist and scatter

OriolAbril commented on src/arviz_plots/plots/ppcplot.py:
```diff
@@ -0,0 +1,529 @@
+        Defaults to empty list. A warning is raised if `pc_kwargs` is also used to define
+        dims to facet over with the `cols` key and `pc_kwargs` takes precedence.
+    data_pairs : dict, optional
+        Dictionary containing relations between observed data and posterior/prior predictive data.
+        Dictionary structure:
+            * key = observed data var_name
+            * value = posterior/prior predictive var_name
+        For example, data_pairs = {'y' : 'y_hat'}.
+        If None, it will assume that the observed data and the posterior/prior predictive data
+        have the same variable name
```
```suggestion
        Dictionary containing relations between observed data and posterior/prior predictive data.
        Dictionary keys are variable names corresponding to observed data and dictionary values
        are variable names corresponding to posterior/prior predictive. For example, data_pairs = {'y' : 'y_hat'}.
        By default, it will assume that the observed data and the posterior/prior predictive data
        have the same variable names
```

OriolAbril commented on src/arviz_plots/plots/ppcplot.py:
```diff
@@ -0,0 +1,529 @@
+        For example, data_pairs = {'y' : 'y_hat'}.
+        If None, it will assume that the observed data and the posterior/prior predictive data
+        have the same variable name
+    aggregate: bool, optional
+        If True, predictive data will be aggregated over both sample_dims and reduce_dims.
+        Defaults to False.
```
```suggestion
    aggregate: bool, default False
        If True, predictive data will be aggregated over both sample_dims and reduce_dims.
```

When the default is the value that is used it should be documented here directly. If it is a placeholder or if it depends on other values then use `optional` and explain the behaviour in the description

OriolAbril commented on src/arviz_plots/plots/ppcplot.py:
```diff
@@ -0,0 +1,529 @@
+        How to represent the marginal density. Defaults to ``rcParams["plot.density_kind"]``.
+    facet_dims : list, optional
+        Dimensions to facet over (for which multiple plots will be generated).
+        Defaults to empty list. A warning is raised if `pc_kwargs` is also used to define
+        dims to facet over with the `cols` key and `pc_kwargs` takes precedence.
```
```suggestion
        Defaults to empty list.
```

OriolAbril commented on src/arviz_plots/plots/ppcplot.py:
```diff
@@ -0,0 +1,529 @@
+    ):
+        raise TypeError(f"`num_pp_samples` must be an integer between 1 and {total_pp_samples}.")
+
+    pp_sample_ix = rng.choice(total_pp_samples, size=num_pp_samples, replace=False)
+
+    # print(f"\npp_sample_ix: {pp_sample_ix!r}")
+
+    # stacking sample dims into a new 'ppc_dim' dimension
+    pp_distribution = pp_distribution.stack(ppc_dim=sample_dims)
+
+    # Select the desired samples
+    pp_distribution = pp_distribution.isel(ppc_dim=pp_sample_ix)
+
+    # renaming sample_dims so that rest of plot will consider this as sample_dims
+    sample_dims = ["ppc_dim"]
```
I think we should skip all these steps if num_pp_samples == total_pp_samples. I think nothing within the function itself should change (test probably will) and it would allow using aesthetics for chain/draw when plotting all samples

OriolAbril replied:
we might also want to distinguish between length 1 sample_dims (in which case it is not necessary to stack nor to rename and overwrite its value) and >1 length sample_dims (in which case stacking is necessary)

OriolAbril commented on src/arviz_plots/plots/ppcplot.py:
```diff
@@ -0,0 +1,529 @@
+            kind=kind,
+            plot_collection=plot_collection,
+            labeller=labeller,
+            aes_map={
+                kind: value for key, value in aes_map.items() if key == "predictive"
+            },  # aes_map[kind] is set to "predictive" aes_map
```
```suggestion
            aes_map={kind: aes_map.get("predictive", {})},
```

Not sure if you were getting errors when "predictive" key wasn't in `aes_map` but this should not be a loop, we are always selecting only one key and we know which one beforehand. Same for the other similar situations.

OriolAbril commented on src/arviz_plots/visuals/__init__.py:
```diff
@@ -59,9 +60,13 @@ def trace_rug(da, target, backend, mask, xname=None, y=None, **kwargs):
             xvalues = da[xname]
         if y is None:
             y = da.min().item()
+    if flatten is not False:
+        xvalues = xvalues.values.flatten()  # flatten xvalues
     if len(xvalues.shape) != 1:
         raise ValueError(f"Expected unidimensional data but got {xvalues.sizes}")
```
can you try what happens when xvalues is always flattened and no error is risen?

OriolAbril commented on src/arviz_plots/plots/ppcplot.py:
```diff
@@ -0,0 +1,529 @@
+    # checking if plot_kwargs["observed"] or plot_kwargs["aggregate"] is not inconsistent with
+    # top level bool arguments `observed`` and `aggregate`
+
+    # observed args logic check:
+    # observed will be True/False depending on user-input/group- true for posterior, false for prior
+    # if observed and no plot_kwargs['observed'], no prob
+    # if observed = True and plot_kwargs['observed'] = True, no prob (observed is plotted)
+    # if observed = True and plot_kwargs['observed'] = False, prob (error/warning raised)
+    # if observed = False and plot_kwargs['observed'] = True, prob (error/warning raised)
+    # if observed = False and plot_kwargs['observed'] = False, no prob (observed is not plotted)
+    if (
+        observed
+        and plot_kwargs.get("observed", True) is False
+        or not observed
+        and plot_kwargs.get("observed", False) is not False
+    ):
+        # raise warning or error
+        raise ValueError(
+            f"""
+            `observed` and `plot_kwargs["observed"]` inconsistency detected.
+            `observed` = {observed}{default_observed}
+            `plot_kwargs["observed"]` = {plot_kwargs["observed"]}
+            Please make sure `observed` and `plot_kwargs["observed"]` have the same value."""
+        )
+
+    # same check for aggregate:
+    if (
+        aggregate
+        and plot_kwargs.get("aggregate", True) is False
+        or not aggregate
+        and plot_kwargs.get("aggregate", False) is not False
+    ):
+        # raise warning or error
+        raise ValueError(
+            f"""
+            `aggregate` and `plot_kwargs["aggregate"]` inconsistency detected.
+            `aggregate` = {aggregate}
+            `plot_kwargs["aggregate"]` = {plot_kwargs["aggregate"]}
+            Please make sure `aggregate` and `plot_kwargs["aggregate"]` have the same value."""
+        )
```
Seeing the tests and having reviewed plot_forest, I realized this is different behaviour from `plot_forest`. It might make sense to mimic the behaviour completely, which would be raising an error if `plot_kwargs["aggregate"]` (or observed) is `False`, continue on ahead otherwise.

This would mean we could have a dict in that key that gets ignored and the element is not plotted, but that is also what happens in plot_forest and I think is good enough behaviour.

OriolAbril commented on tests/test_hypothesis_plots.py:
```diff
@@ -172,3 +178,43 @@ def test_plot_ridge(datatree, combined, plot_kwargs, labels_shade_label):
+        },
+    ),
+    kind=ppc_kind_value,
+    group=ppc_group,
```
All relevant plot specific arguments should be here, especially if they affect what gets plotted or how the data is processed _within the function_. Thus, `observed` and `aggregate` (toggling `observed_rug` on too if `observed` is true), `facet_dims` and `num_pp_samples` should be here.

OriolAbril commented on tests/test_plots.py:
```diff
@@ -246,3 +272,52 @@ def test_plot_ridge_aes_labels_shading(self, backend, datatree_4d, pseudo_dim):
             assert all(0 in child["alpha"] for child in pc.aes.children.values())
             assert any(pseudo_dim in child["shade"].dims for child in pc.viz.children.values())
+
+    @pytest.mark.parametrize("group", ("prior", "posterior"))
```
No need to parametrize by group. It plays an extremely minor role in the function and it is also being checked in the hypothesis tests.

OriolAbril commented on tests/test_plots.py:
```diff
@@ -246,3 +272,52 @@ def test_plot_ridge_aes_labels_shading(self, backend, datatree_4d, pseudo_dim):
+    @pytest.mark.parametrize("group", ("prior", "posterior"))
+    @pytest.mark.parametrize("kind", ("kde", "cumulative"))
+    @pytest.mark.parametrize("facet_dims", (["group"], ["hierarchy"], None))
+    def test_plot_ppc_4d(self, datatree_4d, facet_dims, kind, group, backend):
```
there should be assertions to check `facet_dims` is working as expected. For example, making sure that dimension is present in `pc.viz["obs"]["plot"]`

imperorrp commented:
- Updated plot_ppc with changes: docstring updates, a check to only flatten and stack if `num_pp_samples` isn't `total_pp_samples` or `len(sample_dims)`>1, plot_forest-like checks for `plot_kwargs['observed']` and `plot_kwargs['aggregate']` to ensure they aren't False.

- The `trace_rug` visual element now flattens by default 

- Parametrizations for `group` are removed from test_plots, parametrizations for `facet_dims` are added and assert checks for  `facet-dims` in `pc.viz["obs"]["plot"].dims` were added. Since the flattening/stacking and updating of sample_dims to ppc_dims was made conditional on some logic now, the 
`assert "ppc_dim" in pc.viz["obs"].dims`
statement doesn't work anymore- should we keep it with some logic added? 

- Also test_hypothesis_plots now has `observed`, `observed_rug`, `aggregate`, `facet_dims`, `sample_dims`, and a composite strategy for `num_pp_samples` (dependent on draws from `group` and `sample_dims`) added and `plot_kwargs["observed"]` and `plot_kwargs["aggregate"]` are always set to True. 

Modified the `datatree` fixture in test_hypothesis_plots too to name the first two dims in each variable so they can be referenced by `sample_dims` later

There are still 2 failing conditions in the hypothesis tests though -I'm not sure why yet- where the datatree seems to be interpreted as a function for some reason

OriolAbril commented:
>  Since the flattening/stacking and updating of sample_dims to ppc_dims was made conditional on some logic now, the
assert "ppc_dim" in pc.viz["obs"].dims
statement doesn't work anymore- should we keep it with some logic added?

Some checks on this should still be present

OriolAbril commented on tests/test_hypothesis_plots.py:
```diff
@@ -18,13 +18,23 @@ def datatree(seed=31):
+            "posterior_predictive": {"obs": posterior_predictive},
+        },
+        dims={
+            "theta": ["chain", "draw", "hierarchy", "group"],
+            "tau": ["chain", "draw", "hierarchy"],
+            "obs": ["chain", "draw", "hierarchy", "group"],
```
adding chain and draw is actually breaking the tests. `from_dict` assumes the `sample_dims` rcparams prepend the provided dimensions _in some of the groups_, in others, where it doesn't make sense for these dimensions to be they are not prepended. When you say `obs: chain, draw, hierarchy, group` you are breaking this behaviour, for the posterior predictive group it is not really an issue because if they are already present arviz won't prepend them again, but in observed_data they are not present at all, yet the code now says they should

OriolAbril commented on tests/test_hypothesis_plots.py:
```diff
@@ -172,3 +182,86 @@ def test_plot_ridge(datatree, combined, plot_kwargs, labels_shade_label):
+        optional={
+            "kind": plot_kwargs_value,
+            "predictive": plot_kwargs_value,
+            "observed": plot_kwargs_value,
+            "aggregate": plot_kwargs_value,
+            "observed_rug": plot_kwargs_value,
```
I think none of these can be false, so they should not have their values be `plot_kwargs_value`, see what other tests like plot_forest are doing

OriolAbril commented on src/arviz_plots/plots/ppcplot.py:
```diff
@@ -0,0 +1,500 @@
+    ):
+        raise TypeError(f"`num_pp_samples` must be an integer between 1 and {total_pp_samples}.")
+
+    # stacking sample dimensions and selecting randomly if sample_dims length>1 or
+    # num_pp_samples=total_pp_samples
+    if num_pp_samples != total_pp_samples or len(sample_dims) == 1:
+        pp_sample_ix = rng.choice(total_pp_samples, size=num_pp_samples, replace=False)
+
+        # stacking sample dims into a new 'ppc_dim' dimension
+        pp_distribution = pp_distribution.stack(ppc_dim=sample_dims)
+
+        # Select the desired samples
+        pp_distribution = pp_distribution.isel(ppc_dim=pp_sample_ix)
+
+        # renaming sample_dims so that rest of plot will consider this as sample_dims
+        sample_dims = ["ppc_dim"]
```
This is not the logic we discussed.

If `num_pp_samples` is already `total_pp_samples` there is no need to subset (so no rng.choice nor .isel). Consequently, it is not necessary to stack anyhting either.

If len(sample_dims)=1, it might still be necessary to subset (this will depend on whether num_pp_samples is the same than the length of the provided dimension). But there is no need to stack anything, instead, the provided dimension should be maintained and used for subsetting.

imperorrp commented:
Resolved latest reviews with these last commits- the predictive values stacking/subselection logic was updated and the the tests modified. The Hypothesis tests still seem to be failing for 2 conditions though

imperorrp commented:
Just rebased this PR too

OriolAbril commented on tests/test_hypothesis_plots.py:
```diff
@@ -192,3 +202,91 @@ def test_plot_ridge(datatree, combined, plot_kwargs, labels_shade_label):
+    sample_dims=ppc_sample_dims,
+    drawed_samples=draw_num_pp_samples(ppc_group, ppc_sample_dims),
+)
+def test_plot_ppc(
```
```suggestion
def test_plot_ppc(
    datatree,
```

imperorrp commented:
Added support for kind="scatter" type plots in `plot_ppc`. 

- Some work on the 'y' aesthetic that is mapped to all of the predictive rugs needs to be done
- When `observed_rug=True` and `observed=True` for kind='scatter', maybe turning one off would be good because they essentially have the same function 

```
azp.plot_ppc(
        data,
        kind="scatter",
        num_pp_samples=10,
        aggregate=True,
        observed_rug=True,
    )
```
![image](https://github.com/user-attachments/assets/76d4e271-20ee-42bf-bfbb-706784a5d1a7)


```
azp.plot_ppc(
        data,
        kind="scatter",
        num_pp_samples=5,
        aggregate=True,
    )
```
![image](https://github.com/user-attachments/assets/6517f615-86f6-4520-b82a-96a8bb036b19)

imperorrp commented:
Fixed some more test related issues (new errors still persisting though). The 'alpha' keyword was missing from the 'none' backend for the 'hist' function, unlike the matplotlib and bokeh backends, which is why it hadn't been detected yet. 

Another fixed test fail was due to the fact that different artists have different dimensions (for example, the 'predictive' artists are overlayed along sample dims but the 'observed' artist does not have these dimensions and the 'aggregate' artist has these reduced instead of overlayed) so the dimensions were varying when the 'predictive' key in `plot_kwargs` was being turned off. 

Suggestion by @OriolAbril to deal with the above is to check dimensions of the artists in the tests. Current issue with this though is since a lot of the artists `plot_ppc` creates are done so via internal calls to `plot_dist`, they are named based on their 'kind' and thus can have the same name despite different plot-related visual concerns. 

Current tests bypass this if 'predictive' is False in `plot_kwargs`

#### Todo: 

- Renaming of the artists to match `plot_ppc` `plot_kwargs` keys
- Modifying the hypothesis tests to check dimensions of each of these artists 
- Fix new test errors

aloctavodia commented:
Closing in favor of #138
